pkg/base-dev/lib/sss.hoon

```diff
@@ -0,0 +1,280 @@
+    ?~  ;;((soft path) path.res)  ~|  %need-path  !!  
+    ?@  payload.res
+      (pine-response res)
+    (scry-response res)
+  ::
+  ++  pine-response
+    |=  res=[path=paths from=dude =aeon =what]
+    ^-  (quip card:agent:gall _sub)
+    =*  current  [src.bowl from.res path.res]
+    =/  =flow  (~(gut by sub) current *flow)
+    :_  (~(put by sub) current flow(aeon (max aeon.flow aeon.res)))
+    ?-    what.res
+        %rock
+      ?.  |((lth aeon.rok.flow aeon.res) =(aeon.res 0))  ~
+      :~  :*  %pass   (zoom request/scry/(scot %p src.bowl)^from.res^what.res^(scot %ud aeon.res)^path.res)
+              %agent  [src.bowl from.res]
+              %poke   %sss-to-pub  :-  result-type  ^-  result
+              [path.res dap.bowl %scry %rock aeon.res]
+      ==  ==
+    ::
+        %wave
+      =/  cards=(list card:agent:gall)
+        :~  :*  %pass  (zoom behn/(scot %p src.bowl)^from.res^path.res)
+                %arvo  %b  %wait  (add ~s10 now.bowl)
+        ==  ==
+      =?  cards  (gth aeon.res +(aeon.flow))  [(pine %rock current) cards]
+      =?  cards  (gth aeon.res aeon.rok.flow)
+        %+  weld  cards
+        %+  turn  (gulf +(aeon.rok.flow) aeon.res)
+        |=  =aeon
+        ^-  card:agent:gall
+        :*  %pass   (zoom request/scry/(scot %p src.bowl)^from.res^what.res^(scot %ud aeon)^path.res)
+            %agent  [src.bowl from.res]
+            %poke   %sss-to-pub  :-  result-type  ^-  result
+            [path.res dap.bowl %scry %wave aeon]
+        ==
+      cards
+    ==
+  ::
+  ++  scry-response
+    |=  $:  path=paths
+            =dude
+            =aeon
+            $%([what=%rock =rock:lake] [what=%wave =wave:lake])
+        ==
+    ^-  (quip card:agent:gall _sub)
+    =*  current  [src.bowl dude path]
+    =/  =flow  (~(gut by sub) current *flow)
+    ?.  (lth aeon.rok.flow aeon)
+      %.  `sub
+      (slog leaf/"ignoring stale {<what>} at aeon {<aeon>}" ~)
+    |^
+    ?-    what
+        %rock
+      =.  wav.flow  (lot:wav wav.flow `aeon ~)
+      =.  rok.flow  [aeon | rock]
+      =.  aeon.flow  (max aeon aeon.flow)
+      (swim ~)
+    ::
+        %wave
+      ?:  =(aeon +(aeon.rok.flow))
+        =.  rok.flow  [aeon | (wash:lake rock.rok.flow wave)]
+        (swim `wave)
+      `(~(put by sub) current flow(wav (put:wav wav.flow aeon wave)))
+    ==
+    ++  swim
+      |=  wave=(unit wave:lake)
+      ^-  (quip card:agent:gall _sub)
+      =^  wave  wav.flow  (del:wav wav.flow +(aeon.rok.flow))
+      ?^  wave
+        =.  rok.flow  [+(aeon.rok.flow) | (wash:lake rock.rok.flow u.wave)]
+        (swim wave)
+      :_  (~(put by sub) current flow)
+      :~  :*  %pass   (zoom on-rock/(scot %ud aeon.rok.flow)^(scot %p src.bowl)^dude^path)
+              %agent  [our dap]:bowl
+              %poke   %sss-on-rock  :-  on-rock-type  ^-  from
+              [path src.bowl dude rock.rok.flow ^wave]
+      ==  ==
+    --
+  --
+++  du
+  |*  [=(lake) paths=mold]
+  =>
+    |%
+    +$  rule  [rocks=_1 waves=_5]
```

Comment:
I can't figure this out -- what is being tracked here with the constants 1 and 5?